url0d38a10ee7_diegopliebana_GEBT_tgz-pJ8-ch_idsia_scenarios_test_EvolveWithChangingSeedsJ8/src/ch/idsia/mario/engine/sprites/Mario.java

```diff
@@ -0,0 +1,815 @@
+        }
+        if (xa > 0)
+        {
+            sliding = true;
+            if (isBlocking(x + xa + width, y + ya - height, xa, ya)) collide = true;
+            else sliding = false;
+            if (isBlocking(x + xa + width, y + ya - height / 2, xa, ya)) collide = true;
+            else sliding = false;
+            if (isBlocking(x + xa + width, y + ya, xa, ya)) collide = true;
+            else sliding = false;
+        }
+        if (xa < 0)
+        {
+            sliding = true;
+            if (isBlocking(x + xa - width, y + ya - height, xa, ya)) collide = true;
+            else sliding = false;
+            if (isBlocking(x + xa - width, y + ya - height / 2, xa, ya)) collide = true;
+            else sliding = false;
+            if (isBlocking(x + xa - width, y + ya, xa, ya)) collide = true;
+            else sliding = false;
+        }
+
+        if (collide)
+        {
+            if (xa < 0)
+            {
+                x = (int) ((x - width) / 16) * 16 + width;
+                this.xa = 0;
+            }
+            if (xa > 0)
+            {
+                x = (int) ((x + width) / 16 + 1) * 16 - width - 1;
+                this.xa = 0;
+            }
+            if (ya < 0)
+            {
+                y = (int) ((y - height) / 16) * 16 + height;
+                jumpTime = 0;
+                this.ya = 0;
+            }
+            if (ya > 0)
+            {
+                y = (int) ((y - 1) / 16 + 1) * 16 - 1;
+                onGround = true;
+            }
+            return false;
+        }
+        else
+        {
+            x += xa;
+            y += ya;
+            return true;
+        }
+    }
+
+    private boolean isBlocking(float _x, float _y, float xa, float ya)
+    {
+        int x = (int) (_x / 16);
+        int y = (int) (_y / 16);
+        if (x == (int) (this.x / 16) && y == (int) (this.y / 16)) return false;
+
+        boolean blocking = world.level.isBlocking(x, y, xa, ya);
+
+        byte block = world.level.getBlock(x, y);
+
+        if (((Level.TILE_BEHAVIORS[block & 0xff]) & Level.BIT_PICKUPABLE) > 0)
+        {
+            Mario.getCoin();
+            world.level.setBlock(x, y, (byte) 0);
+            for (int xx = 0; xx < 2; xx++)
+                for (int yy = 0; yy < 2; yy++)
+                    world.addSprite(new Sparkle(x * 16 + xx * 8 + (int) (Math.random() * 8), y * 16 + yy * 8 + (int) (Math.random() * 8), 0, 0, 0, 2, 5));
+        }
+
+        if (blocking && ya < 0)
+        {
+            world.bump(x, y, large);
+        }
+
+        return blocking;
+    }
+
+    public void stomp(Enemy enemy)
+    {
+        if (deathTime > 0 || world.paused) return;
+
+        float targetY = enemy.y - enemy.height / 2;
+        move(0, targetY - y);
+
+        xJumpSpeed = 0;
+        yJumpSpeed = -1.9f;
+        jumpTime = 8;
+        ya = jumpTime * yJumpSpeed;
+        onGround = false;
+        sliding = false;
+        invulnerableTime = 1;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 618:629, 755:766.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.